### ISSUE TITLE
val/miou_best metric goes wrong when training with multiple GPU

### ISSUE THREAD

**Issue by Shanci-Li** [drprojects/superpoint_transformer#165]
Hi Damien,

Thanks for your great work!

![image](https://github.com/user-attachments/assets/b6e9afe9-c9e7-489d-86e6-e5dce720202c)

Unfortunately, there seems to be a bug after you have changed some code in batch data processing. The val/miou and val/imou_best are not consistent when training with multiple GPUs and the latter is wrong. Could you have a look and fix this issue? 

Hope you have a good day!

Best regards,
Shanci

**Comment by drprojects** [comment]
Hi, thanks for the info.

We have not tested our code on multi-GPU. Although lightning allows doing so easily, I have not paid particular attention to the distribution normalization layers or metrics computation on multiple devices. So there is no garantee that this works off-the-shelf.

You seem to suggest that you have been doing multi-GPU with the project before and that it worked. Can you confirm ? 

Besides, your message also seems to suggest that things worked fine on multi-gpu and now they don't after a change ? Can you please indicate the breaking commit ? 

Finally, after answering the above questions, if you want us to have a look, please share a minimum reproducible example.

**Comment by Shanci-Li** [comment]
Hi Damien,

Thanks for your considerate reply! I had indeed worked on 2 commits d940e8bc4be8a79afb0ab0d5255693ef2c5f1df8 and cc6b7684fb09f4851d881ba8cf7c7e4330f7ddf0. I double-checked all the training logs and found the issue had occurred in both of the commits. But for the multi GPU training, sometimes the val/miou_best works fine while sometimes it is wrong. That is why I have thought it worked in one commit before... I have not figured out what is the problem and how to reproduce the issue stably. I will try to investigate the reason and update once I find the reason.

Thanks again!

**Comment by drprojects** [comment]
I am not sure I understand: are you saying you observed some discrepancy between `val/miou` and `val/miou` in the single GPU case also or only when using multi-GPU ? Beyond the metrics computation, have you not experienced any other issue with multi-GPU training ? 

In any case, looking forward to your minimum reproducible example.

**Comment by Shanci-Li** [comment]
To clarify, all the metrics with single GPU training/validation/testing are correct while 3 metrics (val/macc_best, val/miou_best, val/oa_best) with multi-GPU training are wrong. The val/macc, val/miou and val/oa are correct based on per class metric in the validation stage, while the va/best_xxx metrics have increased value than they should.

**Comment by drprojects** [comment]
I see, thanks for the clarification ! 

I hope you can find a way to make the error reproducible so we can check. 

In the meantime, after a quick look at the [lightning documentation for distributed metrics computation](https://pytorch-lightning.readthedocs.io/en/1.5.10/advanced/multi_gpu.html), I kinda feel that the problem _might_ be tackled by setting wherever necessary: 

```python
self.log(..., sync_dist=are_we_doing_multi_gpu)
```

Could you please give this a try and see if the issue persists ?

**Comment by Shanci-Li** [comment]
Hi Damien,

After process-level debugging, I found it is exactly the synchronization of distributed training causes the error. In the current code, the val/test set is distributed to processes, and the metrics are calculated on each process independently, while the best metric (best_miou/best_oa etc.) is the highest one among the partitions. Therefore, the metrics in the validation stage are not correct as they are calculated from only part of the validation set.

For reproduction, I think you can simply modify the number of devices > 1 in `/configs/trainer/gpu.yaml`:

```
defaults:
  - default.yaml

accelerator: gpu
devices: 4

# mixed precision for extra speed-up
# precision: 16
# precision: bf16
precision: 32
````
and run the training on the kiiti360 dataset. To observe the above mentioned situation, you can set a break point [here](https://github.com/drprojects/superpoint_transformer/blob/fd7f765a104d772e11757f78ea31506a765662ba/src/models/semantic.py#L689) and check the `self.val_cm.confmat.sum()`, `miou` in the debug consloe.

As you mentioned, the torch lighting provides the `sync_dist` function to sync between processes. However, by default, the reduce_fx is to average the metric from all processes, while in superpoint clusters, each sample distributed to processes does not contribute to the final metric identically. The processed nag is with different number of nodes, which correspond to different number of points in the cloud. Therefore, the confusion matrix from different processes should be aggregated first. Then we can get the right metric based on the updated confmat.

For the implementation, the lighting mentioned in [tutorial](https://pytorch-lightning.readthedocs.io/en/1.5.10/advanced/multi_gpu.html#synchronize-validation-and-test-logging) that 

> Note if you use any built in metrics or custom metrics that use [TorchMetrics](https://torchmetrics.readthedocs.io/), these do not need to be updated and are automatically handled for you.

I am not sure if this means the lighting and TorchMetric could do element-wise addition on ConfusionMatrix() across the processes as we want. SPT did not record the confmat except for the last test epoch and applying the reduce_fx on scalar metric like acc/miou result from non-uniform partition of the validation set does not make sense as well. So I used the second solution in the tutorial. With `self.all_gather` to collect `self.val_cm.confmat` across processes and only log the metric on the rank_zero device, the metric looks correct now.

I do not know if you want to spend time on the TorchMetric.ConfusionMatrix synchronization and whether using `self.all_gather` is enough. I can make a PR if you think my solution is acceptable. :)

**Comment by drprojects** [comment]
Hi @Shanci-Li, thank you for the in-depth investigation and feedback ! Yes, please create a PR so that I better understand how you implemented the confusion matrix aggregation across multiple processes 😉 

If we go that way, the aggregation of panoptic segmentation metrics will also require careful aggregation.

**Comment by drprojects** [comment]
Hi @Shanci-Li, have you made any progress on a potential PR to address this issue you found ?

**Comment by drprojects** [comment]
Thanks for taking the time to do so ! I gave you my feedback in the PR
